File: typescript/src/external_task_worker.ts
/* eslint-disable @typescript-eslint/member-naming */
import {Logger} from 'loggerhythm';
import * as uuid from 'uuid';

import {HttpClient} from '@essential-projects/http';
import {IIdentity} from '@essential-projects/iam_contracts';
import {
  DataModels,
  HandleExternalTaskAction,
  IExternalTaskWorker,
} from '@process-engine/consumer_api_contracts';

import {ExternalTaskApiExternalAccessor} from './accessors/external_accessor';
import {ExternalTaskApiClientService} from './external_task_api_client_service';

const logger: Logger = Logger.createLogger('processengine:external_task:worker');

export class ExternalTaskWorker<TExternalTaskPayload, TResultPayload> implements IExternalTaskWorker {

  private readonly _workerId = uuid.v4();
  private readonly lockDuration = 30000;
  private readonly processEngineUrl: string;
  private readonly topic: string;
  private readonly maxTasks: number;
  private readonly longpollingTimeout: number;
  private readonly processingFunction: HandleExternalTaskAction<TExternalTaskPayload, TResultPayload>;

  private _identity: IIdentity;
  private _pollingActive: boolean = false;
  private externalTaskClient: ExternalTaskApiClientService;

  constructor(
    processEngineUrl: string,
    identity: IIdentity,
    topic: string,
    maxTasks: number,
    longpollingTimeout: number,
    processingFunction: HandleExternalTaskAction<TExternalTaskPayload, TResultPayload>,
  ) {
    this.processEngineUrl = processEngineUrl;
    this.identity = identity;
    this.topic = topic;
    this.maxTasks = maxTasks;
    this.longpollingTimeout = longpollingTimeout;
    this.processingFunction = processingFunction;

    this.initialize();
  }

  public get identity(): IIdentity {
    return this._identity;
  }

  public set identity(value: IIdentity) {
    this._identity = value;
  }

  public get workerId(): string {
    return this._workerId;
  }

  public get pollingIsActive(): boolean {
    return this._pollingActive;
  }

  public start(): void {
    this._pollingActive = true;
    this.processExternalTasks();
  }

  public stop(): void {
    this._pollingActive = false;
  }

  private initialize(): void {
    const httpClient = new HttpClient();
    httpClient.config = {url: this.processEngineUrl};

    const externalAccessor = new ExternalTaskApiExternalAccessor(httpClient);
    this.externalTaskClient = new ExternalTaskApiClientService(externalAccessor);
  }

  private async processExternalTasks(): Promise<void> {

    while (this.pollingIsActive) {

      const externalTasks = await this.fetchAndLockExternalTasks(
        this.identity,
        this.topic,
        this.maxTasks,
        this.longpollingTimeout,
      );

      if (externalTasks.length === 0) {
        await this.sleep(1000);
        continue;
      }

      const executeTaskPromises: Array<Promise<void>> = [];

      for (const externalTask of externalTasks) {
        executeTaskPromises.push(this.executeExternalTask(this.identity, externalTask));
      }

      await Promise.all(executeTaskPromises);
    }
  }

  private async fetchAndLockExternalTasks(
    identity: IIdentity,
    topic: string,
    maxTasks: number,
    longpollingTimeout: number,
  ): Promise<Array<DataModels.ExternalTask.ExternalTask<TExternalTaskPayload>>> {

    try {
      return await this
        .externalTaskClient
        .fetchAndLockExternalTasks<TExternalTaskPayload>(identity, this.workerId, topic, maxTasks, longpollingTimeout, this.lockDuration);
    } catch (error) {

      logger.error(
        'An error occured during fetchAndLock!',
        error.message,
        error.stack,
      );

      // Returning an empty Array here, since "waitForAndHandle" already implements a timeout, in case no tasks are available for processing.
      // No need to do that twice.
      return [];
    }
  }

  private async executeExternalTask(
    identity: IIdentity,
    externalTask: DataModels.ExternalTask.ExternalTask<TExternalTaskPayload>,
  ): Promise<void> {

    try {
      const lockExtensionBuffer = 5000;

      const interval =
        setInterval(async (): Promise<void> => this.extendLocks(identity, externalTask), this.lockDuration - lockExtensionBuffer);

      const result = await this.processingFunction(externalTask);
      clearInterval(interval);

      await this.processResult(identity, result, externalTask.id);

    } catch (error) {
      logger.error('Failed to execute ExternalTask!', error.message, error.stack);
      await this.externalTaskClient.handleServiceError(identity, this.workerId, externalTask.id, error.message, '');
    }
  }

  private async extendLocks(identity: IIdentity, externalTask: DataModels.ExternalTask.ExternalTask<TExternalTaskPayload>): Promise<void> {
    try {
      await this.externalTaskClient.extendLock(identity, this.workerId, externalTask.id, this.lockDuration);
    } catch (error) {
      // This can happen, if the lock-extension was performed after the task was already finished.
      // Since this isn't really an error, a warning suffices here.
      logger.warn(`An error occured while trying to extend the lock for ExternalTask ${externalTask.id}`, error.message, error.stack);
    }
  }

  private async processResult(identity: IIdentity, result: DataModels.ExternalTask.ExternalTaskResultBase, externalTaskId: string): Promise<void> {

    if (result instanceof DataModels.ExternalTask.ExternalTaskBpmnError) {
      const bpmnError = result as DataModels.ExternalTask.ExternalTaskBpmnError;
      await this.externalTaskClient.handleBpmnError(identity, this.workerId, externalTaskId, bpmnError.errorCode);

    } else if (result instanceof DataModels.ExternalTask.ExternalTaskServiceError) {

      const serviceError = result as DataModels.ExternalTask.ExternalTaskServiceError;
      await this
        .externalTaskClient
        .handleServiceError(identity, this.workerId, externalTaskId, serviceError.errorMessage, serviceError.errorDetails);

    } else {
      await this
        .externalTaskClient
        .finishExternalTask(identity, this.workerId, externalTaskId, (result as DataModels.ExternalTask.ExternalTaskSuccessResult).result);
    }
  }

  private async sleep(milliseconds: number): Promise<void> {
    return new Promise<void>((resolve: Function): void => {
      setTimeout((): void => { resolve(); }, milliseconds);
    });
  }

}

File: dotnet/src/ExternalTaskWorker.cs
namespace ProcessEngine.ExternalTaskAPI.Client
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Net.Http;
    using System.Threading;
    using System.Threading.Tasks;

    using EssentialProjects.IAM.Contracts;

    using ProcessEngine.ConsumerAPI.Contracts;
    using ProcessEngine.ConsumerAPI.Contracts.DataModel;

    /// <summary>
    /// Periodically fetches, locks and processes ExternalTasks for a given topic.
    /// </summary>
    public class ExternalTaskWorker<TExternalTaskPayload, TResultPayload> : IExternalTaskWorker
            where TExternalTaskPayload : new()
    {
        private const int LockDuration = 30000;
        private readonly string ProcessEngineUrl;
        private readonly string Topic;
        private readonly int MaxTasks;
        private readonly int LongpollingTimeout;
        private readonly HandleExternalTaskAction<TExternalTaskPayload, TResultPayload> ProcessingFunction;

        private ExternalTaskApiClientService ExternalTaskClient;

        /// <summary>
        /// Creates an new instance of ExternalTaskWorker
        /// </summary>
        public ExternalTaskWorker(
            string processEngineUrl,
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout,
            HandleExternalTaskAction<TExternalTaskPayload, TResultPayload> processingFunction
        )
        {
            this.ProcessEngineUrl = processEngineUrl;
            this.Identity = identity;
            this.Topic = topic;
            this.MaxTasks = maxTasks;
            this.LongpollingTimeout = longpollingTimeout;
            this.ProcessingFunction = processingFunction;

            this.Initialize();
        }

        /// <summary>
        /// The identity the worker should use to poll for ExternalTasks.
        /// </summary>
        /// <value></value>
        public IIdentity Identity { get; set;}

        /// <summary>
        /// Indicates, if the worker is currently polling for ExternalTasks.
        /// </summary>
        public bool PollingActive { get; private set;} = false;

        /// <summary>
        /// The ID of the worker
        /// </summary>
        public string WorkerId { get; } = Guid.NewGuid().ToString();

        public void Start() {
            this.PollingActive = true;
            this.ProcessExternalTasks();
        }

        public void Stop() {
            this.PollingActive = false;
        }

        private void Initialize() {
          var httpClient = new HttpClient();
          httpClient.BaseAddress = new Uri(ProcessEngineUrl);

          this.ExternalTaskClient = new ExternalTaskApiClientService(httpClient);
        }

        /// <summary>
        /// Periodically fetches, locks and processes available ExternalTasks with a given topic,
        /// using the given callback as a processing function.
        /// </summary>
        private async Task ProcessExternalTasks()
        {
            while (this.PollingActive)
            {
                var externalTasks = await this.FetchAndLockExternalTasks(
                    this.Identity,
                    this.Topic,
                    this.MaxTasks,
                    this.LongpollingTimeout
                );

                if (externalTasks.Count() == 0)
                {
                    Thread.Sleep(1000);
                    continue;
                }

                var tasks = new List<Task>();

                foreach (var externalTask in externalTasks)
                {
                    tasks.Add(this.ExecuteExternalTask(Identity, externalTask));
                }

                await Task.WhenAll(tasks);
            }
        }

        private async Task<IEnumerable<ExternalTask<TExternalTaskPayload>>> FetchAndLockExternalTasks(
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout
        )
        {
            try
            {
                return await this.ExternalTaskClient.FetchAndLockExternalTasks<TExternalTaskPayload>(
                  this.Identity,
                  this.WorkerId,
                  this.Topic,
                  this.MaxTasks,
                  this.LongpollingTimeout,
                  LockDuration
                );
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);

                // Returning an empty Array here, since "waitForAndHandle" already implements a timeout, in case no tasks are available for processing.
                // No need to do that twice.
                return new List<ExternalTask<TExternalTaskPayload>>();
            }
        }

        private async Task ExecuteExternalTask(
          IIdentity identity,
          ExternalTask<TExternalTaskPayload> externalTask
        )
        {
            const int lockExtensionBuffer = 5000;
            const int lockRefreshIntervalInMs = LockDuration - lockExtensionBuffer;

            var lockRefreshTimer = this.StartExtendLockTimer(identity, externalTask, lockRefreshIntervalInMs);

            try
            {
                var result = await this.ProcessingFunction(externalTask);

                lockRefreshTimer.Stop();

                await this.ProcessResult(identity, result, externalTask.Id);
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
                lockRefreshTimer.Stop();
                await this.ExternalTaskClient.HandleServiceError(identity, this.WorkerId, externalTask.Id, exception.Message, exception.StackTrace);
            }
        }

        private System.Timers.Timer StartExtendLockTimer(IIdentity identity, ExternalTask<TExternalTaskPayload> externalTask, int intervall)
        {
            var timer = new System.Timers.Timer(intervall);
            timer.Elapsed += async (sender, e) => await ExtendLock(identity, externalTask);
            timer.Start();

            return timer;
        }

        private async Task ExtendLock(IIdentity identity, ExternalTask<TExternalTaskPayload> externalTask)
        {
            try
            {
                await this.ExternalTaskClient.ExtendLock(identity, this.WorkerId, externalTask.Id, LockDuration);
            }
            catch (Exception error)
            {
                // This can happen, if the lock-extension was performed after the task was already finished.
                // Since this isn't really an error, a warning suffices here.
                Console.WriteLine($"An error occured while trying to extend the lock for ExternalTask ${externalTask.Id}", error.Message, error.StackTrace);
            }
        }

        private async Task ProcessResult(IIdentity identity, ExternalTaskResultBase result, string externalTaskId) {

            if (result is ExternalTaskBpmnError) {

                var bpmnError = result as ExternalTaskBpmnError;
                await this.ExternalTaskClient.HandleBpmnError(identity, this.WorkerId, externalTaskId, bpmnError.ErrorCode);

            }
            else if (result is ExternalTaskServiceError<object>)
            {

                var serviceError = result as ExternalTaskServiceError<object>;
                await this
                    .ExternalTaskClient
                    .HandleServiceError(identity, this.WorkerId, externalTaskId, serviceError.ErrorMessage, serviceError.ErrorDetails as string);

            }
            else
            {
                await this.ExternalTaskClient.FinishExternalTask(identity, this.WorkerId, externalTaskId, (result as ExternalTaskSuccessResult<object>).Result);
            }
        }
    }
}
